src/types/unsolicited_response.rs

```diff
@@ -0,0 +1,309 @@
+                flags
+                    .into_iter()
+                    .map(|s| Flag::from(s.to_string()))
+                    .collect(),
+            )),
+            Response::MailboxData(MailboxDatum::Exists(n)) => Ok(UnsolicitedResponse::Exists(n)),
+            Response::MailboxData(MailboxDatum::MetadataUnsolicited { mailbox, values }) => {
+                Ok(UnsolicitedResponse::Metadata {
+                    mailbox: mailbox.to_string(),
+                    metadata_entries: values.iter().map(|s| s.to_string()).collect(),
+                })
+            }
+            Response::Expunge(n) => Ok(UnsolicitedResponse::Expunge(n)),
+            Response::Vanished { earlier, uids } => {
+                Ok(UnsolicitedResponse::Vanished { earlier, uids })
+            }
+            Response::Data {
+                status: Status::Ok,
+                ref code,
+                ref information,
+            } => {
+                let info = information.as_ref().map(|s| s.to_string());
+                if let Some(code) = code {
+                    match ResponseCode::try_from(code) {
+                        Ok(owncode) => Ok(UnsolicitedResponse::Ok {
+                            code: Some(owncode),
+                            information: info,
+                        }),
+                        _ => Err(response),
+                    }
+                } else {
+                    Ok(UnsolicitedResponse::Ok {
+                        code: None,
+                        information: info,
+                    })
+                }
+            }
+            Response::Data {
+                status: Status::Bye,
+                ref code,
+                ref information,
+            } => {
+                let info = information.as_ref().map(|s| s.to_string());
+                if let Some(code) = code {
+                    match ResponseCode::try_from(code) {
+                        Ok(owncode) => Ok(UnsolicitedResponse::Bye {
+                            code: Some(owncode),
+                            information: info,
+                        }),
+                        _ => Err(response),
+                    }
+                } else {
+                    Ok(UnsolicitedResponse::Bye {
+                        code: None,
+                        information: info,
+                    })
+                }
+            }
+            Response::Fetch(id, ref attributes) => {
+                match AttributeValue::try_from_imap_proto_vec(attributes) {
+                    Ok(attrs) => Ok(UnsolicitedResponse::Fetch {
+                        id,
+                        attributes: attrs,
+                    }),
+                    _ => Err(response),
+                }
+            }
+            _ => Err(response),
+        }
+    }
+}
+
+/// Owned version of ResponseCode that wraps a subset of [`imap_proto::ResponseCode`]
```

Comment:
Do we actually need this subset enum? Does the set of possible valid response codes vary between solicited and unsolicited OKs?

Reply:
The immediate reason this exists is to avoid exposing the `imap_proto::ResponseCode` in our public API.

In practice, there are a bunch of possible response codes that will never show up in an unsolicited response, like `BadCharset` and [`Capabilities`](https://github.com/djc/tokio-imap/blob/bb553c9512393c1a040e19194f95595958f0cb32/imap-proto/src/types.rs#L66), because these codes are specifically in response to a request. I would actually say that _most_ of the ResponseCodes will never show up in an unsolicited response, but it is less effort and more complete to just turn the `imap_proto::ResponseCode` into the owned version if we're willing to expose it in out public API.

I think it is less work for us to not wrap this, but this is really about whether or not we want to expose the `imap_proto` types directly. I am fine either way.

Reply:
I'm torn because by having a wrapper we also need to make sure we remember to add to this over time as new response codes are handled by `imap_proto`. But I also agree with you that it probably makes sense to only expose the possible subset here. Happy to defer to your preference here.

Reply:
Well, if we want to keep up with `imap_proto`, then the easiest way is to not wrap it at all, but this same argument eventually applies to `imap_proto::Response`. :-) 

For me this is really about whether or not we want to republish the `imap_proto` types. If we do, then we probably shouldn't wrap `ResponseCode` or `AttributeValue`. We already republish `StatusAttribute`, so if that's fine then it's probably fine to republish the others. It's just that in these cases the user will have to hypothetically deal with more variants than they will practically see, but that's fine and they already have to use a `_` arm because those types are non-exhaustive.

Reply:
I think we're unlikely to want to commit fully to having no `imap_proto` types in our public interface, so I think the measuring stick should be "do we gain something else useful by wrapping". For example, in the `UnsolicitedResponse` case it allows us to list (and thus document) _specifically_ the variants users can expect to come unsolicited. If it's a matter of one or two extra variants, it's not too bad, but half of them are impossible, wrapping seems worthwhile. It's definitely not an exact science.

Reply:
Since it is easier to implement, easier to maintain, and also uses the `into_owned` bits that are obviously there to be used in `imap_proto`, I have changed this to just use the owned versions of `ResponseCode` and `AttributeValue` from `imap_proto`. :)